export AOTI_TORCH_EXPORT on Windows.

xuhancn opened the pull request:
Fixes #139954

reproduce UT:
```cmd
pytest test/inductor/test_torchinductor_codegen_dynamic_shapes.py -k test_device_assert_dynamic_shapes_cpu
```
Issue:
<img width="856" alt="image" src="https://github.com/user-attachments/assets/5fc501a9-54e5-45ac-9fb3-509ec11a7abe">

After fixing:
![Image](https://github.com/user-attachments/assets/883846fb-8e92-4b9c-9400-daab32382a3a)

Reland:
1. Declare export on Windows explicitly.
2. Support cpu, cuda and xpu devices.

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

xuhancn commented:
@pytorchbot rebase

desertfire commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -40,12 +40,11 @@
+// PyTorch2 doesn't currently work on Windows. But, we still need to export aoti
+// functions. If we didn't export these functions, it will cause the inductor
+// failed, please issue: https://github.com/pytorch/pytorch/issues/139954
+// TODO: We can only show message when aoti runtime, but not compiling time.
```
Curious what does "show message" mean here?

xuhancn replied:
@desertfire show “AOTI not supported on Windows.”

xuhancn commented:
@pytorchmergebot merge

izaitsevfb commented:
Looks like this change breaks builds at meta:
```
lld-link: error: duplicate symbol: aoti_torch_delete_tensor_object
>>> defined at xplat\caffe2\torch\csrc\inductor\aoti_torch\shim_common.cpp:216
>>>            libtorch_lib_ovrsource.pic.lib(shim_common.cpp.pic.o)
>>> defined at libCaffe2Backend_gpu.dll
```

xuhancn commented:
> aoti_torch_delete_tensor_object

I will take a look.

izaitsevfb commented:
> > aoti_torch_delete_tensor_object
> 
> I will take a look.

To add more details, the symbol clash happens in a specific context: DLL + binary. Apparently an original warning on the definition is still relevant:
```
// PyTorch2 doesn't currently work on Windows. Exporting these APIs can lead
// to symbol clashes at link time if libtorch is included in a DLL and binary
// that depends on the DLL. As a short term fix, we don't export the symbols.
// In the long term, this will need to be addressed when Windows is supported.
// #define AOTI_TORCH_EXPORT __declspec(dllexport)
#define AOTI_TORCH_EXPORT
#else // !_WIN32
#define AOTI_TORCH_EXPORT
#endif // _WIN32
```

xuhancn committed:
enable export.

refactor AOTI_TORCH_EXPORT macro

add EXPORT_AOTI_FUNCTIONS to cuda/xpu.

add EXPORT_AOTI_FUNCTIONS to torch_python

define EXPORT_AOTI_FUNCTIONS in global scope.

xuhancn commented:
@pytorchmergebot rebase

xuhancn commented:
hi @izaitsevfb , 
On Windows, we need to explicit declarate for export functions. So, I made some change for current code. Could you please help me import to Meta internal environment and validate it?

xuhancn commented:
@pytorchmergebot merge

xuhancn commented:
@pytorchmergebot revert -m "need to fix on XPU."

xuhancn commented:
@pytorchmergebot revert -m="need to fix on XPU." -c="ghfirst"

xuhancn commented:
<img width="854" alt="image" src="https://github.com/user-attachments/assets/b7d7f410-ea32-491a-aa65-669af7f7f189">
Failed screen shot on XPU.

xuhancn commented:
@pytorchmergebot rebase

xuhancn commented:
@pytorchmergebot rebase

xuhancn commented:
@pytorchmergebot merge -f "ignore un-related issues."

malfet commented:
@pytorchmergebot revert -m "This broke sm89 builds" -c ignoredsignal

malfet commented:
> @pytorchmergebot merge -f "ignore un-related issues."

Why do you think there were unrelated? See https://hud.pytorch.org/hud/pytorch/pytorch/39482907bea370da65d04b4d0a2dd39c090bf2ec/1?per_page=50&name_filter=sm89&mergeLF=true this PR clearly caused regression

xuhancn commented:
@pytorchmergebot rebase

xuhancn commented:
> > @pytorchmergebot merge -f "ignore un-related issues."
> 
> Why do you think there were unrelated? See https://hud.pytorch.org/hud/pytorch/pytorch/39482907bea370da65d04b4d0a2dd39c090bf2ec/1?per_page=50&name_filter=sm89&mergeLF=true this PR clearly caused regression

I will take a look, thanks.

xuhancn commented:
@pytorchmergebot rebase -b main

xuhancn commented:
@pytorchmergebot drci

xuhancn commented:
@pytorchmergebot merge

atalman commented:
@pytorchmergebot revert -c ghfirst -m "Break internal tests see errors like: csrc\inductor\aoti_torch\shim_common.cpp(481): error C2491: 'aoti_torch__embedding_bag': definition of dllimport function not allowed"

xuhancn commented:
> @pytorchmergebot revert -c ghfirst -m "Break internal tests see errors like: csrc\inductor\aoti_torch\shim_common.cpp(481): error C2491: 'aoti_torch__embedding_bag': definition of dllimport function not allowed"

Hi @atalman , could you please show me the more detaied information? which module report this issue?

xuhancn commented:
@pytorchmergebot drci

xuhancn commented:
@pytorchmergebot merge -f "skip XPU UTs"

clee2000 commented:
@pytorchbot revert -m "still has failures internally when building, D66923759" -c ghfirst

Error logs look like 
```
    torch\csrc\inductor\aoti_torch\shim_common.cpp(481): error C2491: 'aoti_torch__embedding_bag': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(516): error C2491: 'aoti_torch__fft_c2c': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(619): error C2491: 'aoti_torch__scaled_dot_product_efficient_attention': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(656): error C2491: 'aoti_torch_convolution': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(830): error C2491: 'aoti_torch_copy_': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(853): error C2491: 'aoti_torch__mm_plus_mm_out': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(1041): error C2491: 'aoti_torch_view_as_real': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(1051): error C2491: 'aoti_torch_view_dtype': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(1063): error C2491: 'aoti_torch_save_tensor_handle': definition of dllimport function not allowed
    torch\csrc\inductor\aoti_torch\shim_common.cpp(1097): error C2491: 'aoti_torch_print_tensor_handle': definition of dllimport function not allowed
```

xuhancn commented:
> @pytorchbot revert -m "still has failures internally when building, D66923759" -c ghfirst
> 
> Error logs look like
> 
> ```
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(481): error C2491: 'aoti_torch__embedding_bag': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(516): error C2491: 'aoti_torch__fft_c2c': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(619): error C2491: 'aoti_torch__scaled_dot_product_efficient_attention': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(656): error C2491: 'aoti_torch_convolution': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(830): error C2491: 'aoti_torch_copy_': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(853): error C2491: 'aoti_torch__mm_plus_mm_out': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(1041): error C2491: 'aoti_torch_view_as_real': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(1051): error C2491: 'aoti_torch_view_dtype': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(1063): error C2491: 'aoti_torch_save_tensor_handle': definition of dllimport function not allowed
>     torch\csrc\inductor\aoti_torch\shim_common.cpp(1097): error C2491: 'aoti_torch_print_tensor_handle': definition of dllimport function not allowed
> ```

Hi @clee2000 It was failed at Inductor UT? how to reproduce it?

desertfire commented:
> Hi @clee2000 It was failed at Inductor UT? how to reproduce it?

@xuhancn , https://github.com/pytorch/pytorch/pull/142500/ will fix it. I am landing that one first.

albanD commented on caffe2/CMakeLists.txt:
```diff
@@ -779,6 +779,7 @@ endif()
 
 if(NOT BUILD_LIBTORCHLESS)
 add_library(torch_cpu ${Caffe2_CPU_SRCS})
+# target_compile_definitions(torch_cpu PRIVATE -DEXPORT_AOTI_FUNCTIONS)
```
I guess all the changes here and in the file below are leftovers and should be removed?

xuhancn replied:
@albanD sure, this PR was reverted many times, I'm not sure how to pass meta internal CI, so I keep the old code here.
I will clean up code certainly landed.

malfet replied:
It would be nice if you could get those removed before trying to land the change...

malfet commented:
> @xuhancn , #142500 will fix it. I am landing that one first.

Well, if goal is to hide AOTI on Windows, then yes, it will fix it, but it's likely to break OSS Linux builds, unless we've accidentally dropped `-fvisibility=hidden`...

desertfire commented:
> 
> Well, if goal is to hide AOTI on Windows, then yes, it will fix it, but it's likely to break OSS Linux builds, unless we've accidentally dropped `-fvisibility=hidden`...

But those functions are already declared with `AOTI_TORCH_EXPORT` in `shim.h`.

xuhancn commented:
@pytorchmergebot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but my first attempt to fix internal build does not fix all the cases, so let us try again' -c ghfirst

Some builds are ok, but some fails with linking error:

```
lld-link: error: duplicate symbol: aoti_torch_delete_tensor_object
>>> defined at xplat\caffe2\torch\csrc\inductor\aoti_torch\shim_common.cpp:236
>>>            libtorch_lib_ovrsource.pic.lib(shim_common.cpp.pic.o)
>>> defined at libCaffe2Backend_cpu.dll
```

xuhancn commented:
> @pytorchbot revert -m 'Sorry for reverting your change, but my first attempt to fix internal build does not fix all the cases, so let us try again' -c ghfirst
> 
> Some builds are ok, but some fails with linking error:
> 
> ```
> lld-link: error: duplicate symbol: aoti_torch_delete_tensor_object
> >>> defined at xplat\caffe2\torch\csrc\inductor\aoti_torch\shim_common.cpp:236
> >>>            libtorch_lib_ovrsource.pic.lib(shim_common.cpp.pic.o)
> >>> defined at libCaffe2Backend_cpu.dll
> ```

@desertfire still failed in `fbcode`.

xuhancn committed:
try to fix aoti_torch_delete_tensor_object

xuhancn commented:
@pytorchmergebot rebase -b main

xuhancn commented:
@pytorchmergebot rebase -b main

xuhancn commented:
Hi @desertfire I checked PR history, `8e96efc48f6f734ba125974c4111b0c0a50eb360` may fix the `aoti_torch_delete_tensor_object` issue, could you please help on import code and test it?

xuhancn commented:
Will land after https://github.com/pytorch/pytorch/pull/144081

xuhancn commented:
@pytorchmergebot merge

clee2000 commented:
@pytorchbot revert -m "still failing internally D67556174, see D67866123 for link to error" -c ghfirst
It's still
```
lld-link: error: duplicate symbol: aoti_torch_delete_tensor_object
>>> defined at xplat\caffe2\torch\csrc\inductor\aoti_torch\shim_common.cpp:236
>>>            libtorch_lib_ovrsource.pic.lib(shim_common.cpp.pic.o)
>>> defined at libCaffe2Backend_gpu.dll
```

@desertfire could you take another look at this?

malfet commented on CMakeLists.txt:
```diff
@@ -1101,6 +1101,9 @@ if(NOT MSVC)
   endif()
 else()
+  # Define export functions for AOTI.
+  add_compile_definitions(EXPORT_AOTI_FUNCTIONS)
```
So looks like the problem is very simple, someone needs to add the same define to respective bzl file internally, otherwise it will not work

huydhn replied:
For more context, I attempted to do that before the holiday in https://www.internalfb.com/diff/D67525962?dst_version_fbid=937729108340657&transaction_fbid=435395002838074, but I probably did it wrong or missed other places so there were still some build failures left [T211170671](https://www.internalfb.com/intern/tasks/?t=211170671).  I didn't have time to follow up on the issue at that time, so I did a revert instead.

malfet replied:
After some debugging, apparently there is a case when someone tries to load two instances of libtorch (one linked as shared library and another statically) into the same address space, and this causes change to the internal build system (that @huydhn proposed) to fail....

As AOTI is not used in internal Windows deployments, in short term I propose just disable it, but in the long run someone needs to modify buck to mark two targets incompatible with each other

malfet replied:
But I don't understand from PR description, why this define needs to be here

malfet replied:
@xuhancn why this change needs to be there? I.e. shouldn't it be a private define of say `torch_cpu`?

malfet reviewed (changes requested):
Requesting changes as someone needs to make corresponding build changes internally. Will have meeting with @clee2000 to look into build system changes. @desertfire do you want to join?

malfet commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -44,8 +44,11 @@
 // In the long term, this will need to be addressed when Windows is supported.
-// #define AOTI_TORCH_EXPORT __declspec(dllexport)
-#define AOTI_TORCH_EXPORT
+#ifdef EXPORT_AOTI_FUNCTIONS
+#define AOTI_TORCH_EXPORT __declspec(dllexport)
+#else
+#define AOTI_TORCH_EXPORT __declspec(dllimport)
+#endif
```
```suggestion
#ifdef OVRSOURCE
// Do not export AOTI on Windows for internal builds
#define AOTI_TORCH_EXPORT
#ifdef EXPORT_AOTI_FUNCTIONS
#define AOTI_TORCH_EXPORT __declspec(dllexport)
#else
#define AOTI_TORCH_EXPORT __declspec(dllimport)
#endif
#endif /* OVRSOURCE */
```

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -1049,6 +1051,7 @@ if(USE_XPU)
   add_library(torch_xpu ${Caffe2_XPU_SRCS})
   torch_compile_options(torch_xpu)  # see cmake/public/utils.cmake
   target_compile_definitions(torch_xpu PRIVATE USE_XPU)
+  # target_compile_definitions(torch_xpu PRIVATE -DEXPORT_AOTI_FUNCTIONS)
```
```suggestion
```

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -958,6 +959,7 @@ elseif(USE_CUDA)
   set(CUDA_LINK_LIBRARIES_KEYWORD)
   torch_compile_options(torch_cuda)  # see cmake/public/utils.cmake
   target_compile_definitions(torch_cuda PRIVATE USE_CUDA)
+  # target_compile_definitions(torch_cuda PRIVATE -DEXPORT_AOTI_FUNCTIONS)
```
```suggestion
```

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -779,6 +779,7 @@ endif()
 
 if(NOT BUILD_LIBTORCHLESS)
 add_library(torch_cpu ${Caffe2_CPU_SRCS})
+# target_compile_definitions(torch_cpu PRIVATE -DEXPORT_AOTI_FUNCTIONS)
```
```suggestion
```

malfet commented on torch/CMakeLists.txt:
```diff
@@ -310,6 +310,7 @@ endif()
 
 add_library(torch_python SHARED ${TORCH_PYTHON_SRCS})
 torch_compile_options(torch_python)  # see cmake/public/utils.cmake
+target_compile_definitions(torch_python PRIVATE -DEXPORT_AOTI_FUNCTIONS)
```
Why this is needed?
```suggestion
```

malfet committed:
Apply suggestions from code review

xuhancn commented:
@pytorchmergebot rebase -b main

xuhancn committed:
add missed #else for #ifdef OVRSOURCE.

xuhancn commented:
Hi @malfet Thanks for your review and suggestion. Based on your suggestion, I found we still need to fix build issue by https://github.com/pytorch/pytorch/pull/140030/commits/41b398234c97df952414b3c688b8a6806cec71eb 
Please review it agsin.

xuhancn committed:
remove define EXPORT_AOTI_FUNCTIONS for each modules.

xuhancn commented:
@pytorchmergebot merge

xuhancn commented:
@pytorchmergebot merge -f "skip fbcode check, pre meta suggested code."
